only calculate subsurface for nitrogen

emlys opened the pull request:
# Description
Fixes #494 

* Removed inputs `subsurface_critical_length_p` and `subsurface_eff_p`. These were used along with distance to stream to calculate subsurface NDR for phosphorus (`sub_ndr_p.tif`).
* Removed output `sub_ndr_p.tif`. This was calculated legitimately, but IMO it's confusing to include it.
* Removed output `sub_load_p.tif`. This was always filled with 0s.
* Changed the `export` outputs to clarify their meaning:
    * Removed `n_export.tif`, which was the total nitrogen export (surface and subsurface)
    * Removed `p_export.tif`, which was the total phosphorus export (surface and subsurface, though subsurface was always 0)
    * Added `n_surface_export.tif`: surface nitrogen export
    * Added `n_subsurface_export.tif`: subsurface nitrogen export
    * Added `n_total_export.tif`: total nitrogen export (equivalent to the former `n_export.tif`)
    * Added `p_surface_export.tif`: surface phosphorus export (equivalent to the former `p_export.tif`)
* Changed the aggregate vector fields to correspond to the new export outputs.
* Since there are now more field names in the aggregate vector, it was difficult to make them unique and understandable in just 10 characters. I changed the aggregate vector to gpkg format so that longer field names are allowed. 
* Update to modern string formatting syntax

I'm working on the corresponding user's guide changes in a branch off main.

# Checklist
- [x] Updated HISTORY.rst (if these changes are user-facing)
- [ ] Updated the user's guide (if needed)

emlys commented on src/natcap/invest/ndr/ndr.py:
```diff
@@ -204,18 +181,10 @@
     's_bar_path': 's_bar.tif',
     's_factor_inverse_path': 's_factor_inverse.tif',
     'stream_path': 'stream.tif',
-    'sub_crit_len_n_path': 'sub_crit_len_n.tif',
-    'sub_crit_len_p_path': 'sub_crit_len_p.tif',
-    'sub_eff_n_path': 'sub_eff_n.tif',
-    'sub_eff_p_path': 'sub_eff_p.tif',
-    'sub_effective_retention_n_path': 'sub_effective_retention_n.tif',
-    'sub_effective_retention_p_path': 'sub_effective_retention_p.tif',
```
These files weren't used at all.

emlys commented on tests/test_ndr.py:
```diff
@@ -314,66 +309,11 @@ def test_validation(self):
         self.assertEqual(set(invalid_args), set(expected_missing_args))
-
-    @staticmethod
-    def _assert_regression_results_equal(
```
this wasn't used anywhere

emlys commented on src/natcap/invest/ndr/ndr.py:
```diff
@@ -405,7 +362,6 @@ def _validate_inputs(nutrients_to_process, lucode_to_parameters):
 
     # these are used for aggregation in the last step
     field_pickle_map = {}
-    field_header_order_list = []
```
This isn't needed since dictionaries now keep their order.

dcdenu4 commented on HISTORY.rst:
```diff
@@ -75,6 +75,36 @@ Unreleased Changes
+      * Renamed the output ``p_export.tif`` to ``p_surface_export.tif`` to
+        clarify that it only models the surface export of phosphorus.
+      * Renamed the output ``n_export.tif`` to ``n_total_export.tif`` to
+        clarify that is is the total of surface and subsurface nitrogen export.
```
```suggestion
        clarify that it is the total of surface and subsurface nitrogen export.
```

dcdenu4 commented on src/natcap/invest/ndr/ndr.py:
```diff
@@ -996,6 +990,42 @@ def _mult_op(*array_nodata_list):
+        raster_path_list (list): list of single band raster paths.
+        target_nodata (float): desired target nodata value.
+        target_result_path (string): path to float 32 target raster
+            multiplied where all rasters are not nodata.
```
Do you mean __sum__ instead of __multiplied__ here?

emlys replied:
Oops, yeah

dcdenu4 commented on src/natcap/invest/ndr/ndr.py:
```diff
@@ -996,6 +990,42 @@ def _mult_op(*array_nodata_list):
+            if nodata is not None:
+                valid_mask &= ~numpy.isclose(array, nodata)
+        result[valid_mask] = array_nodata_list[0][valid_mask]
+        for array in array_nodata_list[2::2]:
```
Passing in the nodata values as raw and indexing this way is a neat implementation!

dcdenu4 replied:
Is there a reason to grab the first array into result instead of iterating with `array_nodata_list[0::2]`?

emlys replied:
I refactored all this - see above

dcdenu4 commented on src/natcap/invest/ndr/ndr.py:
```diff
@@ -996,6 +990,42 @@ def _mult_op(*array_nodata_list):
+        result = numpy.empty(array_nodata_list[0].shape)
+        result[:] = target_nodata
+        valid_mask = numpy.full(result.shape, True)
+        for array, nodata in zip(*[iter(array_nodata_list)]*2):
```
Could you walk me through `*[iter(array_nodata_list)]*2`? I'm having a hard time understanding how it's unpacking with `zip`? Thanks!

emlys replied:
I just copied all this from the similar raster multiplication function in NDR haha. I should've looked more closely - this implementation is pretty confusing. Refactored the whole thing.

dcdenu4 commented on src/natcap/invest/ndr/ndr.py:
```diff
@@ -996,6 +990,42 @@ def _mult_op(*array_nodata_list):
+
+    """
+    def _sum_op(*array_nodata_list):
+        """Multiply non-nodata stacks."""
```
Sum instead of multiply?

dcdenu4 commented on src/natcap/invest/ndr/ndr.py:
```diff
@@ -1366,10 +1380,12 @@ def _aggregate_and_pickle_total(
+
+    base_wkt = pygeoprocessing.get_vector_info(
+        base_vector_path)['projection_wkt']
+    # use reproject_vector to create a copy in geopackage format
+    # keeping the original projection
+    pygeoprocessing.reproject_vector(
+        base_vector_path, base_wkt, target_vector_path, driver_name='GPKG')
```
I'm curious if `CreateCopy` doesn't work for GPKG drivers, or if there was another reason to piggy back off of `reproject_vector` here?

emlys replied:
`CreateCopy` raised an error when I tried to convert shapefile -> gpkg. I suspect it's not really meant for file conversion.

emlys replied:
Would be nice to have a pygeoprocessing function specifically for copying vectors.

dcdenu4 replied:
I believe we did have some functions like that back in a much earlier API implementation. I certainly remember writing a lot of vector copying, clipping, and intersection code when the gdal Python bindings were a lot less robust. When James and Rich refactored that to the modern day pygeoprocessing, I think a lot of that stuff was left behind.

dcdenu4 reviewed (changes requested):
Hey @emlys thanks for jumping on this issue. I don't even remember making it! I had a few comments and questions. It will also make sense to wait for 3.10 to be merged into main.

emlys commented:
@dcdenu4 I addressed your comments and refactored that whole sum function

dcdenu4 reviewed (commented):
Thanks for those changes @emlys. It looks like there are some conflicts because of the 3.10 release merge. I was also wondering if there was a UG update with these changes as well? And going off our conversation last call, did you happen to run this through the UI?

emlys commented:
> I was also wondering if there was a UG update with these changes as well?

I updated the UG commit hash for that (it's on a new `release/3.10.2` branch)

> did you happen to run this through the UI?

Good catch! I removed those inputs and ran NDR through the UI and it worked.

emlys commented:
Thanks for catching that Doug. There was a mix of tabs and spaces in there causing the weird indents.

dcdenu4 reviewed (approved):
Looks good to me!
